dns/dnssecalgs/__init__.py

```diff
@@ -0,0 +1,106 @@
+    (Algorithm.RSASHA256, None): PrivateRSASHA256,
+    (Algorithm.RSASHA512, None): PrivateRSASHA512,
+    (Algorithm.ECDSAP256SHA256, None): PrivateECDSAP256SHA256,
+    (Algorithm.ECDSAP384SHA384, None): PrivateECDSAP384SHA384,
+    (Algorithm.ED25519, None): PrivateED25519,
+    (Algorithm.ED448, None): PrivateED448,
+}
+
+
+def get_algorithm_cls(
+    algorithm: Union[int, str], prefix: Optional[bytes] = None
+) -> Type[GenericPrivateKey]:
+    """Get Private Key class from Algorithm.
+
+    *algorithm*, a ``str`` or ``int`` specifying the DNSKEY algorithm.
+
+    Raises ``UnsupportedAlgorithm`` if the algorithm is unknown.
+
+    Returns a ``dns.dnssecalgs.GenericPrivateKey``
+    """
+    algorithm = Algorithm.make(algorithm)
+    cls = algorithms.get((algorithm, prefix))
+    if cls:
+        return cls
+    raise UnsupportedAlgorithm(
+        'algorithm "%s" not supported by dnspython' % Algorithm.to_text(algorithm)
+    )
+
+
+def get_algorithm_cls_from_dnskey(dnskey: DNSKEY) -> Type[GenericPrivateKey]:
+    """Get Private Key class from DNSKEY.
+
+    *dnskey*, a ``DNSKEY`` to get Algorithm class for.
+
+    Raises ``UnsupportedAlgorithm`` if the algorithm is unknown.
+
+    Returns a ``dns.dnssecalgs.GenericPrivateKey``
+    """
+    prefix = None
+    if dnskey.algorithm == Algorithm.PRIVATEDNS:
+        _, length = dns.name.from_wire(dnskey.key, 0)
+        prefix = dnskey.key[0:length]
+    elif dnskey.algorithm == Algorithm.PRIVATEOID:
+        length = int(dnskey.key[0])
+        prefix = dnskey.key[0 : length + 1]
+    return get_algorithm_cls(dnskey.algorithm, prefix)
+
+
+def register_algorithm_cls(
+    algorithm: Union[int, str],
+    algorithm_cls: Type[GenericPrivateKey],
+    name: Optional[Union[dns.name.Name, str]] = None,
+    oid: Optional[bytes] = None,
+) -> None:
+    """Register Algorithm Private Key class.
+
+    *algorithm*, a ``str`` or ``int`` specifying the DNSKEY algorithm.
+
+    *algorithm_cls*: A `GenericPrivateKey` class.
+
+    *name*, an optional ``dns.name.Name`` or ``str``, for for PRIVATEDNS algorithms.
+
+    *oid*: an optional BER-encoded `bytes` for PRIVATEOID algorithms.
+
+    Raises ``ValueError`` if a name or oid is specified incorrectly.
+    """
+    algorithm = Algorithm.make(algorithm)
+    if not issubclass(algorithm_cls, GenericPrivateKey):
+        raise TypeError("Invalid algorithm class")
+    prefix = None
+    if algorithm == Algorithm.PRIVATEDNS and name:
+        if isinstance(name, str):
+            name = dns.name.from_text(name)
+        prefix = name.to_wire()
+    elif algorithm == Algorithm.PRIVATEOID and oid:
+        prefix = bytes([len(oid)]) + oid
+    else:
+        if name and algorithm != Algorithm.PRIVATEDNS:
+            raise ValueError("Name only supported for PRIVATEDNS algorithm")
+        if oid and algorithm != Algorithm.PRIVATEOID:
+            raise ValueError("OID only supported for PRIVATEOID algorithm")
```

Comment:
It looks like this allows specifying PRIVATEDNS without a name, or PRIVATEOID without an OID.  Is that intentional?  If so, I don't think it's possible to match those.

Reply:
Yes, you can register a generic PRIVATEDNS/PRIVATEOID algorithm without any prefix matching. This was intentional. Perhaps it is featurism?

Reply:
I don't think it actually works, though.  You can register it, which means that it would have a prefix of `None`.  When `get_algorithm_cls_from_dnskey` sees a DNSKEY record with that private type, though, it will attempt to extract a prefix.  If it successfully extracts a prefix, it will not be `None`, and won't match.  If it fails to successfully extract a prefix, it will throw an exception, and not look in the table.

It's possible that a direct call to `get_algorithm_cls` would find it, but that's far less common.

Reply:
Fixed in 44f2b70fd14dcffc3dbc99023c2223e8893efcc1. Not sure we want to allow prefix-less private algorithms though, WDYT?

Reply:
I don't think it's worth supporting them.

Reply:
Removed in 8fb949c19e47aea291b3bca0310d2e3030eb6448

Reply:
I think there are some unnecessary checks left.  In the final `else`, the algorithm will never be PRIVATEDNS or PRIVATEOID, as those are checked above.

Reply:
Fixed in 8fb949c19e47aea291b3bca0310d2e3030eb6448 and 32577f1d4a8c69b1c56a2480ed092d2380a32915